feat(codedeploy): CodeDeploy deployment config constructs for Lambda and ECS

clareliguori opened the pull request:
CloudFormation now supports Lambda and ECS in the `AWS::CodeDeploy::DeploymentConfig` resource type.  This PR adds L2 constructs specific to ECS and Lambda for that resource type.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fc9178b3cc47a73b2f049e30aa4f9c0d09fdd24d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f9beac19-f0c9-4961-a614-eee99dc7af1c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

cplee commented:
Looks great! I proposed [changes](https://github.com/clareliguori/aws-cdk/pull/3) to deprecate the `CustomLambdaDeploymentConfig` and address the tests.

clareliguori commented:
Thanks @cplee!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9214223a3784edfe425b816e32c5e00a70f7e8a4
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fe5e6a18-fa8c-48f9-b093-cc887e253670%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: c1daa00fc8cee092a8059b4390f1ef9666464322
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=25d0445a-bbe6-44ad-82f2-92db044a6a7e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

corymhall commented on packages/@aws-cdk/aws-codedeploy/README.md:
````diff
@@ -327,3 +292,107 @@ const deploymentGroup = codedeploy.LambdaDeploymentGroup.fromLambdaDeploymentGro
+);
+```
+
+## ECS Applications
````
From this README. I can't tell how I would actually deploy an ECS application using these
constructs. Can you add an example showing how you would use these constructs to setup a blue/green
deployment of an actual ECS service?

clareliguori replied:
I will make a follow-up PR for the deployment group L2 construct that fills in these details in the README.  I wanted to keep this PR relatively small.
https://github.com/clareliguori/aws-cdk/commit/7d29f934e72057c4606e4f8438565461ec82ccd0

corymhall commented on packages/@aws-cdk/aws-codedeploy/test/ecs/integ.deployment-config.ts:
```diff
@@ -0,0 +1,19 @@
+const app = new cdk.App();
+const stack = new cdk.Stack(app, 'aws-cdk-codedeploy-ecs-config');
+
+new codedeploy.EcsDeploymentConfig(stack, 'LinearConfig', {
```
Can you update this integ test to perform an actual ecs deployment? I know it will probably require
manual steps, but that will help us ensure that we are testing an actual scenario.

clareliguori replied:
Yes, though that might be more useful in my follow-up PR with an L2 construct for deployment groups.  These changes are dependent on this PR, so I havent put it up yet:
https://github.com/clareliguori/aws-cdk/commit/7d29f934e72057c4606e4f8438565461ec82ccd0

corymhall reviewed (dismissed):
I think overall I would like to see an example of how this would be used to deploy an ECS service.
Since the lambda & ecs changes are very simliar I only commented on the ecs changes, but the
comments also apply to lambda as well.

clareliguori commented on packages/@aws-cdk/aws-codedeploy/lib/ecs/deployment-config.ts:
```diff
@@ -1,46 +1,109 @@
    */
-  public static fromEcsDeploymentConfigName(_scope: Construct, _id: string, ecsDeploymentConfigName: string): IEcsDeploymentConfig {
+  public static fromEcsDeploymentConfigName(scope: Construct, id: string, ecsDeploymentConfigName: string): IEcsDeploymentConfig {
+    ignore(scope);
```
`_scope` fails a linter rule now that `EcsDeploymentConfig` extends Resource:
```
error: [awslint:from-signature:@aws-cdk/aws-codedeploy.EcsDeploymentConfig.fromEcsDeploymentConfigName.params[0]] invalid method signature for fromXxx method.   (expected="scope",actual="_scope")
```

clareliguori commented on packages/@aws-cdk/aws-codedeploy/lib/ecs/deployment-config.ts:
```diff
@@ -1,46 +1,109 @@
+  public static fromEcsDeploymentConfigName(scope: Construct, id: string, ecsDeploymentConfigName: string): IEcsDeploymentConfig {
+    ignore(scope);
+    ignore(id);
     return deploymentConfig(ecsDeploymentConfigName);
```
We can do that for `fromEcsDeploymentConfigName` method, but we still need the `deploymentConfig` method for the static configs like `ALL_AT_ONCE`.  They don't have any stack/scope context.

corymhall replied:
Yeah that's fine to keep it for the static methods. For `fromEcsDeploymentConfigName` using
`Stack.formatArn` would grab the values (region, account, etc) from the stack instead of using the
generic intrinsic values.

clareliguori commented on packages/@aws-cdk/aws-codedeploy/lib/ecs/deployment-config.ts:
```diff
@@ -1,46 +1,109 @@
-export class EcsDeploymentConfig {
+export class EcsDeploymentConfig extends Resource implements IEcsDeploymentConfig {
   public static readonly ALL_AT_ONCE = deploymentConfig('CodeDeployDefault.ECSAllAtOnce');
+  public static readonly LINEAR_10PERCENT_EVERY_1MINUTES = deploymentConfig('CodeDeployDefault.ECSLinear10PercentEvery1Minutes');
```
Today, these are also used when creating a CodePipeline ECS deploy action:
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-codepipeline-actions/lib/codedeploy/ecs-deploy-action.ts#L154-L157

With the future PR for L2 deployment group constructs, they are used when configuring deployment groups.

clareliguori replied:
Add docstrings and updated README

clareliguori commented:
@corymhall I think I addressed all your feedback, expect the pieces that I expect to cover in my next PR (L2 CodeDeploy deployment group constructs for ECS).  If you think it would be easier to review deployment config + groups as a single PR, I can merge those changes into this PR.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 86541b362ca5fce8a0a60e95c0e628eb90b73c9c
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f807ecdf-5564-44ff-b257-d96a5e8b324a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a91098cadfc4f307eab2fa49e470d11dfd94baa0
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8450b1f6-bd41-4159-a8e4-86e69e6ebafb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

corymhall commented on packages/@aws-cdk/aws-codedeploy/lib/traffic-routing-config.ts:
```diff
@@ -0,0 +1,159 @@
+  /**
+   * represents the underlying traffic routing config structure
+   */
+  readonly config: CfnDeploymentConfig.TrafficRoutingConfigProperty;
```
Can you either flatten this out into `TrafficRoutingConfig` or add a new interface that represents
the `TrafficRoutingConfigProperty`? We don't like to leak anything from the L1 layer, even if that
means duplicating some stuff.

corymhall reviewed (dismissed):
> @corymhall I think I addressed all your feedback, expect the pieces that I expect to cover in my next PR (L2 CodeDeploy deployment group constructs for ECS).  If you think it would be easier to review deployment config + groups as a single PR, I can merge those changes into this PR.

ah ok if you have another PR for deployment group that will address that then I am fine with holding
on that stuff.

Just a couple of minor comments and then I think we are good!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3f9eb7f8eeb29e6aa34543572af12ca370f1b92b
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6644c947-e6d6-4353-8f9a-de7c5a8a7dc3%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a0405ce1822f8bcb77c93fea9b4c15fa3fe7f301
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=39f94dd1-917b-4b96-a332-d4ce1e167779%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

corymhall commented on packages/@aws-cdk/aws-codedeploy/lib/ecs/deployment-config.ts:
```diff
@@ -1,46 +1,128 @@
  * @resource AWS::CodeDeploy::DeploymentConfig
  */
-export class EcsDeploymentConfig {
+export class EcsDeploymentConfig extends Resource implements IEcsDeploymentConfig {
```
I think it might also make sense to create a command abstract base class `DeploymentConfigBase`. It looks like almost all of the logic between ECS and Lambda are the same.

clareliguori replied:
Done!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: af4252a3b52363b7f4f2b3dd16c9e0c8a2bf47d2
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=17ecf68d-74ac-4ab3-b4ec-4fe3c59912a9%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a3498f038a05bddfc94a8ac329c5704eefce498a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7c90faa6-34b1-498e-a3af-375de1c183e7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

corymhall commented on packages/@aws-cdk/aws-codedeploy/lib/base-deployment-config.ts:
```diff
@@ -0,0 +1,159 @@
+    });
+
+    // Traffic routing is not applicable to Server-based deployment configs
+    if (props?.trafficRouting && (props?.computePlatform === undefined || props?.computePlatform == ComputePlatform.SERVER)) {
```
```suggestion
    if (props?.trafficRouting && (props?.computePlatform === undefined || props?.computePlatform === ComputePlatform.SERVER)) {
```

corymhall commented on packages/@aws-cdk/aws-codedeploy/lib/base-deployment-config.ts:
```diff
@@ -0,0 +1,159 @@
+   */
+  public readonly deploymentConfigArn: string;
+
+  public constructor(scope: Construct, id: string, props?: BaseDeploymentConfigProps, additionalProps?: any) {
```
Can the props from `additionalProps` be added to `props`? I think that is just the
`minimumHealthyHosts` property from the server deployment config.

clareliguori replied:
Done!

corymhall reviewed (changes requested):
Thanks for your patience on this! Just two minor things and then I think that is it!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 313ecd3e739ac393ea660a392c34d69e42ce9cbf
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=04e37721-cf4d-4a1f-9831-b4bfc0e8882a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

corymhall commented on packages/@aws-cdk/aws-codedeploy/lib/base-deployment-config.ts:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    // Minimum healthy hosts is only applicable to Server-based deployment configs
+    if (props?.minimumHealthyHosts && props?.computePlatform && props?.computePlatform != ComputePlatform.SERVER) {
```
```suggestion
    if (props?.minimumHealthyHosts && props?.computePlatform !== ComputePlatform.SERVER) {
```

clareliguori replied:
Argh I always forget strict comparison 😂

Typescript question: wouldn't your suggestion throw the error if minimumHealthyHosts is defined but computePlatform is not defined?  That's intended to be a valid case and should not throw the error

corymhall replied:
ah ok thats true, so keep the extra `&&` in there.

corymhall reviewed (dismissed):
Last one

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 62a24966024c0692d77c16e57fd40c5c884794c9
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3e941cbf-6f11-487d-9233-0804dcf0fbab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a8112fb7de3b7eb4b45d1bf09f26d0f52a3df60d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=04df316c-bee5-4093-90b2-33df31da431c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
